### ISSUE TITLE
CA2208 is not reported for .ctor(string? paramName, string? message) when exception derives from ArgumentException

### ISSUE THREAD

**Issue by drieseng** [dotnet/roslyn-analyzers#6580]
### Analyzer

**Diagnostic ID**: [CA2208](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2208): `Instantiate argument exceptions correctly'

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0-preview1.23121.1

### Describe the bug

**CA2208** is not reported when an incorrect parameter name is specified in `.ctor(string? paramName, string? message)` of an exception that derives from **ArgumentException**.

### Steps To Reproduce

1. Use the following .editorconfig fragment to configure **CA2208** with severity **error**:
```txt
dotnet_diagnostic.CA2208.severity = error
```
2. Compile the following code:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentNullException("Not name", "Must not be null.");
        }
    }
}
```

### Expected behavior

The folllowing diagnostic is reported:

error CA2208: Method Reproduce passes 'Not name' as the paramName argument to a ArgumentNullException constructor. Replace this argument with one of the method's parameter names. Note that the provided parameter name should have the exact casing as declared on the method.

### Actual behavior

No diagnostic is reported.

**CA2208** is reported just fine if you change the type of the exception to **ArgumentException**:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentException("Must not be null.", "Not name");
        }
    }
}
```

**CA2208** is also reported if you use the `.ctor(string? paramName)` of **ArgumentNullException**:
```c#
using System;

public sealed class InstantiateCorrectly
{
    public void Reproduce(string name)
    {
        if (name is null)
        {
            throw new ArgumentNullException("Not name");
        }
    }
}
```

**Comment by mpidash** [comment]
The analyzer **finds** the diagnostic described in this issue but does not **report** it.

The reason for that is that each argument of an exception is checked in order (line 107). If the check leads to an `RuleIncorrectMessage` diagnostic, then it is reported and the analyzer is done (line 123). Otherwise, the next parameter is checked.

The problem is that previous diagnostics are discarded (line 120) when the next parameter is checked. This leads to the bug described:

1. Check first parameter
2. Result is `RuleIncorrectParameterName` -- so do not instantly report it and continue loop
3. Check second parameter -- overwrites first diagnostic
4. Result is null -- nothing is reported

https://github.com/dotnet/roslyn-analyzers/blob/dfd38ae3897cb4d8496feb9f04e1eaaddf70741b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs#L106-L127

My proposed fix: Save diagnostics for each argument if it is not a `RuleIncorrectMessage` (otherwise just report it and return). Check at the end if there are any diagnostics and return the last one (to not change the behavior as it is right now).

**Comment by mavasani** [comment]
@buyaa-n for guidance

**Comment by buyaa-n** [comment]
> My proposed fix: Save diagnostics for each argument if it is not a RuleIncorrectMessage (otherwise just report it and return). Check at the end if there are any diagnostics and return the last one (to not change the behavior as it is right now).

Thanks for the fix, the solution sounds good except using a list for saving the diagnostics (I don't think there will be a more than one diagnostic ever)
